### PR TITLE
Implement changes to make plugin work with Pipelines

### PR DESCRIPTION
**Pull request by ButkiewiczP**
This PR builds on #486 

There were a few bugs with that PR, but it doesn't seem like anyone is working on it now. The additions in this PR add support for Pipelines, and fix some of the issues in the other PR, such as proper Github status updates.

The second commit simply changes the default template for the jenkins build description and removes the HTML. HTML doesn't render in the build description, so I'm not sure why it's included in the default template.

*Note on Tests:* I'm not sure why tests fail with a NullPointerException. If you drop the Jenkins version below 1.6, all tests pass. Putting the Jenkins version above 1.6 causes the way JSON is deserialized (in the tests) to fail. Before tests are run, we attempt to set up a Trigger object using the line:

```GhprbTrigger trigger = spy(req.bindJSON(GhprbTrigger.class, json));```

But the bindJSON call in Stapler throws an error, and refuses to move past this.

So -- For what it's worth, I've done a bunch of testing locally with this plugin, and it works, but I can't get the tests to pass, so YMMV. If someone can take a look at the way tests are set up and recommend a solution, I'd be more than willing to implement it.

### CODE REVIEW

**Comment by benpatterson** [comment]
Thanks for taking this another step forward, @ButkiewiczP 

Indeed, whenever I've looked at this kind of upgrade myself, I end up hitting the nullpointer exceptions in the `bindJSON` call. Haven't gotten to a solution yet; I wonder if the trigger object never gets created. But thank you for codifying this; hopefully we can figure out what's up with the error and get this enhancement into the plugin.

File: src/test/java/org/jenkinsci/plugins/ghprb/GhprbTestUtil.java [hunk 1]
```diff
@@ -372,11 +362,12 @@ private static void setupReq() throws Exception{
         given(req.bindJSON(any(Class.class), any(Class.class), any(JSONObject.class))).willCallRealMethod();
         given(req.setBindInterceptpr(any(BindInterceptor.class))).willCallRealMethod();
```
Comment (hughsaunders) [inline src/test/java/org/jenkinsci/plugins/ghprb/GhprbTestUtil.java]: setBindIntercept_**o**_r

I spent hours battling this NPE, I knew it was a mocking problem but couldn't see it. Now looking at your PR the typo in setBindinterceptpr is blindingly obvious. I checked out your PR locally and all the tests pass 🎉   :shipit:

**Comment by ButkiewiczP** [comment]
@hughsaunders Huge thanks for spotting that. What an awful typo to have in the code base 😑 .

PR has been updated to include the code to fix all the broken tests. Hopefully this is enough to get this merged in 😄

**Comment by hughsaunders** [comment]
@ButkiewiczP Thanks for your work on this. I'm running with a locally compiled version of this plugin but have a problem where github pull request builder is not saved when saving a job. I add the trigger to a job, configure it, click save, reload the config page for that job and the trigger config is gone :/

logs:
```
Mar 22, 2017 2:29:18 PM WARNING hudson.model.Descriptor verifyNewInstance
Father of org.jenkinsci.plugins.ghprb.GhprbTrigger@1c089694 and its getDescriptor() points to two different instances. Probably malplaced @Extension. See http://hudson.361315.n4.nabble.com/Help-Hint-needed-Post-build-action-doesn-t-stay-activated-td2308833.html
Mar 22, 2017 2:29:18 PM INFO org.jenkinsci.plugins.ghprb.GhprbTrigger stop
Stopping the ghprb trigger for project scratchpipeline
Mar 22, 2017 2:29:18 PM INFO org.jenkinsci.plugins.ghprb.GhprbTrigger start
Starting the ghprb trigger for the scratchpipeline job; newInstance is true
```

**Comment by ButkiewiczP** [comment]
@hughsaunders Thanks for picking up on that. Turned out to be a misplaced `@Extension`,  not exactly sure how that wound up in there. I tested locally before and after pushing up the latest commit, and have solved the issue.

**Comment by hughsaunders** [comment]
Thanks @ButkiewiczP for the quick response, I'll try again with the latest version.

**Comment by hughsaunders** [comment]
@ButkiewiczP works for me, thanks! :shipit:

**Comment by ButkiewiczP** [comment]
@hughsaunders Glad to hear it

@benpatterson Is there anything more I can or should do to move this along? Would really like to get this into the Jenkins repo.

**Comment by benpatterson** [comment]
@ButkiewiczP I'll take a closer look in the coming days. Psyched you found
the typo. Surely that's been there a long time.

File: src/main/java/org/jenkinsci/plugins/ghprb/GhprbTrigger.java [hunk 2]
```diff
@@ -376,7 +358,15 @@ public void run() {
...
-        return this.job.scheduleBuild2(job.getQuietPeriod(), cause, new GhprbParametersAction(values), buildData);
+
+        ParameterizedJobMixIn scheduledJob = new ParameterizedJobMixIn() {
+            @Override
+            protected Job asJob() {
+                return job;
+            }
+        };
+
+        return scheduledJob.scheduleBuild2(Jenkins.getInstance().getQuietPeriod(),new CauseAction(cause),new GhprbParametersAction(values), buildData);
```
Comment (benpatterson) [inline src/main/java/org/jenkinsci/plugins/ghprb/GhprbTrigger.java]: could you move down the comments on lines 358-360?

File: src/test/java/org/jenkinsci/plugins/ghprb/GhprbTestUtil.java [hunk 3]
```diff
@@ -370,13 +360,13 @@ private static void setupReq() throws Exception{
         
         given(req.bindJSON(any(Class.class), any(JSONObject.class))).willCallRealMethod();
         given(req.bindJSON(any(Class.class), any(Class.class), any(JSONObject.class))).willCallRealMethod();
-        given(req.setBindInterceptpr(any(BindInterceptor.class))).willCallRealMethod();
```
Comment (benpatterson) [inline src/test/java/org/jenkinsci/plugins/ghprb/GhprbTestUtil.java]: oh man

File: src/main/java/org/jenkinsci/plugins/ghprb/GhprbPullRequestMerge.java [hunk 4]
```diff
@@ -95,21 +81,21 @@ void setHelper(Ghprb helper) {
     }
 
     @Override
-    public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, final BuildListener listener) throws InterruptedException, IOException {
-        logger = listener.getLogger();
-        AbstractProject<?, ?> project = build.getProject();
-        if (build.getResult().isWorseThan(Result.SUCCESS)) {
-            logger.println("Build did not succeed, merge will not be run");
-            return true;
+    public void perform(@Nonnull Run<?, ?> run, @Nonnull FilePath filePath, @Nonnull Launcher launcher, @Nonnull TaskListener taskListener) throws InterruptedException, IOException {
```
Comment (benpatterson) [inline src/main/java/org/jenkinsci/plugins/ghprb/GhprbPullRequestMerge.java]: Sorry if I'm missing something obvious, but why is a filePath needed, and why can't it be `null`?

File: src/main/java/org/jenkinsci/plugins/ghprb/GhprbPullRequestMerge.java [hunk 5]
```diff
@@ -161,14 +147,14 @@ public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, final Build
 
         // The build should not fail if no merge is expected
         if (intendToMerge && canMerge && (isMergeable == null || !isMergeable)) {
-            logger.println("Pull request cannot be automerged.");
+            listener.getLogger().println("Pull request cannot be automerged.");
             commentOnRequest("Pull request is not mergeable.");
-            listener.finished(Result.FAILURE);
-            return false;
+            listener.error(Result.FAILURE.toString());
```
Comment (ButkiewiczP) [inline src/main/java/org/jenkinsci/plugins/ghprb/GhprbPullRequestMerge.java]: My understanding is `.getLogger()` returns the `PrintStream` object that will receive the output of the build. The `.error()` method returns a `PrintWriter` object that is `A writer to receive details of the error` according to the docs. My guess is Jenkins treats these differently somewhere upstream, but I'm not certain.

File: src/main/java/org/jenkinsci/plugins/ghprb/GhprbPullRequestMerge.java (hunk 4 above)
Comment (ButkiewiczP) [inline src/main/java/org/jenkinsci/plugins/ghprb/GhprbPullRequestMerge.java]: This method is defined in the `SimpleBuildStep` interface which the `GhprbPullRequestMerge` class implements.

**Comment by ButkiewiczP** [comment]
@benpatterson I think my latest commit should address all the issues you mentioned above. Let me know if you have other suggestions.

**Comment by irvingoe** [comment]
@ButkiewiczP sorry, maybe a silly question but will the branc variable (sha1) and the refspec configuration remain the same?
Those configurations were added at the 'Source Code Management' job section but that section is not present in pipeline projects, or is there something I'm missing?

File: src/test/java/org/jenkinsci/plugins/ghprb/GhprbTestUtil.java [hunk 6]
```diff
@@ -370,13 +360,13 @@ private static void setupReq() throws Exception{
...
+        given(req.setBindInterceptor(any(BindInterceptor.class))).willCallRealMethod();
         given(req.setBindListener(any(BindInterceptor.class))).willCallRealMethod();
-        given(req.getBindInterceptor()).willCallRealMethod();
         given(req.getWebApp()).willReturn(webApp);
         
         req.setBindListener(BindInterceptor.NOOP);
         req.setBindInterceptpr(BindInterceptor.NOOP);
```
Comment (jasonkuster) [inline src/test/java/org/jenkinsci/plugins/ghprb/GhprbTestUtil.java]: FYI: Typo here too. :)

File: src/main/java/org/jenkinsci/plugins/ghprb/Ghprb.java [hunk 7]
```diff
@@ -342,16 +333,24 @@ public static GhprbCause getCause(Run<?, ?> build) {
...
 
-    public static GhprbTrigger extractTrigger(AbstractProject<?, ?> p) {
-        GhprbTrigger trigger = p.getTrigger(GhprbTrigger.class);
-        if (trigger == null || (!(trigger instanceof GhprbTrigger))) {
-            return null;
+    public static GhprbTrigger extractTrigger(Job<?, ?> p) {
+
+        ParameterizedJobMixIn.ParameterizedJob pJob = (ParameterizedJobMixIn.ParameterizedJob) p;
+        GhprbTrigger ghprbTrigger = null;
+        if (p instanceof ParameterizedJobMixIn.ParameterizedJob) {
```
Comment (jasonkuster) [inline src/main/java/org/jenkinsci/plugins/ghprb/Ghprb.java]: Isn't this guaranteed to be true based on line 342? I feel like this method might be clearer if it started off with the type check instead. e.g:

```java
public static GhprbTrigger extractTrigger(Job<?, ?> p) {
    if (!(p instanceof ParameterizedJobMixIn.ParameterizedJob)) {
        return null;
    }
    ParameterizedJobMixIn.ParemeterizedJob pjob = (ParameterizedJobMixIn.ParameterizedJob) p;
    ...
}

File: src/main/java/org/jenkinsci/plugins/ghprb/GhprbPullRequestMerge.java [hunk 8]
```diff
@@ -95,21 +84,21 @@ void setHelper(Ghprb helper) {
     }
 
     @Override
-    public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, final BuildListener listener) throws InterruptedException, IOException {
-        logger = listener.getLogger();
```
Comment (jasonkuster) [inline src/main/java/org/jenkinsci/plugins/ghprb/GhprbPullRequestMerge.java]: If you leave this line in you can revert a number of the changes below, for what it's worth.

**Comment by benpatterson** [comment]
Hi - Also checking in here. I'm looking at this on a local environment and am now encountering the crumb issue described [here](https://issues.jenkins-ci.org/browse/JENKINS-12875). There is a fix for another plugin that can be found [here](https://github.com/jan-molak/jenkins-build-monitor-plugin/commit/16e5866374375d57608f77e46abd1eb6db25da8d).

Error message for any future googling is `WARNING: No valid crumb was included in request for /jenkins/ghprbhook/. Returning 403.`

I have not ruled out errors in my set-up, but since a change appears to have been introduced in Jenkins 2, I suspect a change is needed. If I can get to this, I'll let you know...if OTOH you have the time, feel free to take a look....

**Comment by benpatterson** [comment]
Re: the csrf protection, there's a better reference implementation [here](https://github.com/jenkinsci/github-plugin/blob/93d40692ff3866705175624e93ec584d4ac88132/src/main/java/com/cloudbees/jenkins/GitHubWebHookCrumbExclusion.java). I'm obviously looking into it but also don't have a lot of cycles the next couple of weeks so thought I'd call it out in a slow-motion #livetweet

**Comment by ButkiewiczP** [comment]
@benpatterson Can you elaborate on the crumb issue you're seeing? When do you see the error? Are you able to trigger and build pipeline jobs after pushing pull requests at all? Is there any easy way to replicate this?

I only ask because I haven't seen this issue in my local set up.

**Comment by benpatterson** [comment]
@ButkiewiczP If it helps, I have a docker instance running jenkins 2.7.3 (admittedly old). I manually configured a webhook on a test repo I use for this purpose, and I'm using ngrok to capture the webhooks and point them at my instance. This had not been a problem on my docker image before this change, but again it's quite possible I have made a mistake in the setup somewhere. So I say with perhaps 75% certainty that this will be an issue. 😰

**Comment by hughsaunders** [comment]
@benpatterson  @ButkiewiczP I had a crumb related error when I forgot /ghprbhook/ from the hook url. Reproduced just now: 

```
Webhook URL: http://1fbaea60.ngrok.io/
Response:
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8"/>
<title>Error 403 No valid crumb was included in the request</title>
</head>
<body><h2>HTTP ERROR 403</h2>
<p>Problem accessing /. Reason:
<pre>    No valid crumb was included in the request</pre></p><hr><i><small>Powered by Jetty://</small></i><hr/>

</body>
</html>
```
```
Webhook URL: http://1fbaea60.ngrok.io/ghprbhook/
Response: Empty/200
```

**Comment by ButkiewiczP** [comment]
@hughsaunders Thanks for pointing that out.

I haven't been setting up my webhooks manually. 

image

The checkbox to automatically manage the webhook works for me and hasn't given me any crumb errors. @benpatterson Can you double check your hook URL is correct? Or try to automatic hook management? If you still get the error then I'll start digging into a fix.

**Comment by benpatterson** [comment]
Folks indeed as I was worried about, looks like the webhook was user error 😞 

Thanks for working through the various details. We may leave a couple of things for later but I'm planning on merging this soon.

**Comment by benpatterson** [comment]
As noted, I simplified the existing PR a bit and removed the changes to the template.

Very excited how this was a community-driven enhancement. Thanks everyone who commented, coded, and patiently made this work. I'll release ASAP.

**Comment by benpatterson** [comment]
Released in v 1.36.0. It'll take ~24 hours to be available on the cloudbees infrastructure.

**Pull request merged**